server/src/main/java/org/elasticsearch/search/aggregations/pipeline/InternalPercentileRanksBucket.java

```diff
@@ -0,0 +1,202 @@
+                    " one of the computed percentile ranks.  Available keys are: " + Arrays.toString(percents));
+        }
+        return percentile_ranks;
+    }
+
+    @Override
+    public String percentileRankAsString(double percent) {
+        return format.format(percentile_rank(percent)).toString();
+    }
+
+    DocValueFormat formatter() {
+        return format;
+    }
+
+    @Override
+    public Iterator<Percentile> iterator() {
+        return new Iter(percents, percentile_ranks);
+    }
+
+    @Override
+    public double value(String name) {
+        return percentile_rank(Double.parseDouble(name));
+    }
+
+    @Override
+    public InternalMax doReduce(List<InternalAggregation> aggregations, ReduceContext reduceContext) {
+        throw new UnsupportedOperationException("Not supported");
+    }
+
+    @Override
+    public XContentBuilder doXContentBody(XContentBuilder builder, Params params) throws IOException {
+        if (keyed) {
+            builder.startObject("values");
+            for (double percent : percents) {
+                double value = percentile_rank(percent);
+                boolean hasValue = !(Double.isInfinite(value) || Double.isNaN(value));
+                String key = String.valueOf(percent);
+                builder.field(key, hasValue ? value : null);
+                if (hasValue && format != DocValueFormat.RAW) {
+                    builder.field(key + "_as_string", percentileAsString(percent));
+                }
+            }
+            builder.endObject();
+        } else {
+            builder.startArray("values");
+            for (double percent : percents) {
+                double value = percentile_rank(percent);
+                boolean hasValue = !(Double.isInfinite(value) || Double.isNaN(value));
+                builder.startObject();
+                builder.field("key", percent);
+                builder.field("value", hasValue ? value : null);
+                if (hasValue && format != DocValueFormat.RAW) {
+                    builder.field(String.valueOf(percent) + "_as_string", percentileRankAsString(percent));
+                }
+                builder.endObject();
+            }
+            builder.endArray();
+        }
+        return builder;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj) return true;
+        if (obj == null || getClass() != obj.getClass()) return false;
+        if (super.equals(obj) == false) return false;
+
+        InternalPercentileRanksBucket that = (InternalPercentileRanksBucket) obj;
+        return Arrays.equals(percents, that.percents) && Arrays.equals(percentile_ranks, that.percentile_ranks);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(super.hashCode(), Arrays.hashCode(percents), Arrays.hashCode(percentile_ranks));
+    }
+
+    public static class Iter implements Iterator<PercentileRank> {
```

Comment:
I think we can reuse the `Percentile` class that percentiles, percentile_ranks and percentiles_bucket uses.  E.g. here is the TDigest percentile ranks iterator using `Percentile`: https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/search/aggregations/metrics/InternalTDigestPercentileRanks.java#L87

Which means we can remove `PercentileRanks#PercentileRank` above

Reply:
I will take a look on this and try to digest this `TDigest` class over this weekend. :). 
Yep, while writing code, one of thing I wasn't complete sure of was do I need `PercentileRank` class or not. This helps. will revisit this.